src/rust/multichan/template/src/template.rs

```diff
@@ -0,0 +1,391 @@
+					MangaStatus::Completed
+				} else {
+					MangaStatus::Unknown
+				};
+				Manga {
+					id,
+					cover,
+					title,
+					author: String::from(author.trim()),
+					artist: String::new(),
+					description,
+					url,
+					categories,
+					status,
+					nsfw,
+					viewer,
+				}
+			})
+			.collect::<Vec<_>>();
+		Ok(MangaPageResult { manga, has_more })
+	}
+
+	pub fn get_manga_list(&self, filters: Vec<Filter>, page: i32) -> Result<MangaPageResult> {
+		let mut title = String::new();
+		let mut sort = String::from("&n=");
+		let mut order_by_date_when_search = false;
+		let mut tags = Vec::new();
+		for filter in filters {
+			match filter.kind {
+				FilterType::Title => title = urlencode(filter.value.as_string()?.read()),
+				FilterType::Sort => {
+					if let Ok(value) = filter.value.as_object() {
+						let index = value.get("index").as_int().unwrap_or(0);
+						let asc = value.get("ascending").as_bool().unwrap_or(false);
+						sort.push_str(match index {
+							0 => "date",
+							1 => "fav",
+							2 => "abc",
+							3 => "ch",
+							_ => "",
+						});
+						sort.push_str(match asc {
+							true => "asc",
+							false => "desc",
+						});
+						if &sort == "&n=datedesc" {
+							sort = String::new();
+						}
+					}
+				}
+				FilterType::Genre => {
+					if let Ok(id) = filter.object.get("id").as_string() {
+						let id = id.read();
+						match filter.value.as_int().unwrap_or(-1) {
+							0 => tags.push(format!("-{}", id)),
+							1 => tags.push(id),
+							_ => continue,
+						}
+					}
+				}
+				FilterType::Check => {
+					let value = filter.value.as_int().unwrap_or(-1);
+					match filter.name.as_str() {
+						"Сортировка по дате" => {
+							order_by_date_when_search = value == 1
+						}
+						_ => continue,
+					}
+				}
+				_ => continue,
+			}
+		}
+
+		let url = if title.is_empty() && tags.is_empty() {
+			format!(
+				"{url}/manga/new?offset={offset}{sort}",
+				url = self.base_url,
+				offset = (page - 1) * 20,
+				sort = sort,
+			)
+		} else if title.is_empty() {
+			format!(
+				"{url}/tags/{tags}?offset={offset}{sort}",
+				url = self.base_url,
+				tags = tags.join("+"),
+				offset = (page - 1) * 20,
+				sort = sort,
+			)
+		} else {
+			format!(
+				"{url}/?do=search&subaction=search&search_start={page}&full_search=0&result_from={offset}&result_num=40&story={title}{need_sort_date}",
+				url=self.base_url,
+				page=page,
+				offset=(page - 1) * 40 + 1,
+				title=title,
+				need_sort_date=if order_by_date_when_search { "&need_sort_date=true" } else { "" },
+			)
+		};
+		let html = Request::new(&url, HttpMethod::Get).html();
+		self.parse_manga_list(html, if !title.is_empty() { 40 } else { 20 })
+	}
+
+	pub fn get_manga_listing(&self, listing: Listing, _: i32) -> Result<MangaPageResult> {
+		if &listing.name == "Случайная" {
+			let html = Request::new("https://manga-chan.me/manga/random", HttpMethod::Get).html();
```

Comment:
I think this messes up the other two sources because they also have the same listing but you're only reading from the manga-chan website. Do the other sites have the same endpoint? You can probably just use the base_url here.

Reply:
oh oops that was my bad, other sources have the same endpoints too